src/python/pants/rules/core/cloc.py

```diff
@@ -0,0 +1,130 @@
+from pants.engine.isolated_process import ExecuteProcessRequest, ExecuteProcessResult
+from pants.engine.legacy.graph import HydratedTargets, TransitiveHydratedTargets
+from pants.engine.rules import console_rule, rule
+from pants.engine.selectors import Get
+from pants.util.objects import datatype
+
+
+class DownloadedClocScript(datatype([('script_path', str), ('digest', Digest)])):
+  """Cloc script as downloaded from the pantsbuild binaries repo."""
+
+
+#TODO(#8459) - It would be nice if we weren't hardcoding a specific version of the cloc binary here.
+@rule
+def download_cloc_script() -> DownloadedClocScript:
+  url = "https://binaries.pantsbuild.org/bin/cloc/1.80/cloc"
+  sha_256 = "2b23012b1c3c53bd6b9dd43cd6aa75715eed4feb2cb6db56ac3fbbd2dffeac9d"
+  digest = Digest(sha_256, 546279)
+  snapshot = yield Get(Snapshot, UrlToFetch(url, digest))
+  yield DownloadedClocScript(script_path=snapshot.files[0], digest=snapshot.directory_digest)
+
+
+class CountLinesOfCode(Goal):
+  name = 'cloc'
+
+  @classmethod
+  def register_options(cls, register) -> None:
+    super().register_options(register)
+    register('--transitive', type=bool, fingerprint=True, default=True,
+             help='Operate on the transitive dependencies of the specified targets.  '
+                  'Unset to operate only on the specified targets.')
+    register('--ignored', type=bool, fingerprint=True,
+             help='Show information about files ignored by cloc.')
+
+
+@console_rule
+def run_cloc(console: Console, options: CountLinesOfCode.Options, cloc_script: DownloadedClocScript, specs: Specs) -> CountLinesOfCode:
+  """Runs the cloc perl script in an isolated process"""
+
+  transitive = options.values.transitive
+  ignored = options.values.ignored
+
+  if transitive:
+    targets = yield Get(TransitiveHydratedTargets, Specs, specs)
+    all_target_adaptors = {t.adaptor for t in targets.closure}
+  else:
+    targets = yield Get(HydratedTargets, Specs, specs)
+    all_target_adaptors = {t.adaptor for t in targets}
+
+  digests_to_merge = []
+
+  build_root = get_buildroot()
+  source_paths: Set[str] = set()
+  for t in all_target_adaptors:
+    sources = getattr(t, 'sources', None)
+    if sources is not None:
+      digests_to_merge.append(sources.snapshot.directory_digest)
+      for f in sources.snapshot.files:
+        source_paths.add(str(f))
+
+  file_content = '\n'.join(sorted(source_paths)).encode()
+
+  input_files_filename = 'input_files.txt'
+  report_filename = 'report.txt'
+  ignore_filename = 'ignored.txt'
+
+  input_file_list = InputFilesContent(FilesContent((FileContent(path=input_files_filename, content=file_content, is_executable=False),)))
+  input_file_digest = yield Get(Digest, InputFilesContent, input_file_list)
+  cloc_script_digest = cloc_script.digest
+  digests_to_merge.extend([cloc_script_digest, input_file_digest])
+  digest = yield Get(Digest, DirectoriesToMerge(directories=tuple(digests_to_merge)))
+
+  cmd = (
+    '/usr/bin/perl',
```

Comment:
This deserves a comment calling the hack out. It's a hack we also rely on to execute pexes, although we leverage PATH there in the form of the `PythonSetup` `--interpreter-search-paths` option: https://github.com/pantsbuild/pants/blob/455190b28609e33acc1d340dbf6733e7b980294a/pants.remote.ini#L30-L40

Consider filing an issue to track either:
1. Handling bootstrap of implicit binary requirements like these via BinaryUtil, or the more deep bootstrapping of a c compiler toolchain all the way through building perl all via rules.
2. Finding a way to express / impose these requirements on users so they get a better / earlier error message that the environment hosting the EPR doesn't have the required binary.